pydrive2/fs.py

```diff
@@ -0,0 +1,476 @@
+
+    @_gdrive_retry
+    def gdrive_get_file(self, item_id, rpath, callback):
+        param = {"id": item_id}
+        # it does not create a file on the remote
+        gdrive_file = self.client.CreateFile(param)
+
+        extra_args = {}
+        if callback:
+            gdrive_file.FetchMetadata(fields="fileSize")
+            callback.set_size(int(gdrive_file.get("fileSize")))
+            extra_args["callback"] = callback.update
+
+        gdrive_file.GetContentFile(rpath, **extra_args)
+
+    def _open(self, path, mode, **kwargs):
+        assert mode in {"rb", "wb"}
+        if mode == "wb":
+            return GDriveBufferedWriter(self, path)
+        else:
+            item_id = self._get_item_id(path)
+            return self.gdrive_open_file(item_id)
+
+    @_gdrive_retry
+    def gdrive_open_file(self, item_id):
+        param = {"id": item_id}
+        # it does not create a file on the remote
+        gdrive_file = self.client.CreateFile(param)
+        fd = gdrive_file.GetContentIOBuffer()
+        return IterStream(iter(fd))
+
+    def rm_file(self, path):
+        item_id = self._get_item_id(path)
+        self.gdrive_delete_file(item_id)
+
+    @_gdrive_retry
+    def gdrive_delete_file(self, item_id):
+        from pydrive2.files import ApiRequestError
+
+        param = {"id": item_id}
+        # it does not create a file on the remote
+        item = self.client.CreateFile(param)
+
+        try:
+            item.Trash() if self._trash_only else item.Delete()
+        except ApiRequestError as exc:
+            http_error_code = exc.error.get("code", 0)
+            if (
+                http_error_code == 403
+                and self._list_params["corpora"] == "drive"
+                and exc.GetField("location") == "file.permissions"
+            ):
+                raise PermissionError(
+                    "Insufficient permissions to {}. You should have {} "
+                    "access level for the used shared drive. More details "
+                    "at {}.".format(
+                        "move the file into Trash"
+                        if self._trash_only
+                        else "permanently delete the file",
+                        "Manager or Content Manager"
+                        if self._trash_only
+                        else "Manager",
+                        "https://support.google.com/a/answer/7337554",
+                    )
+                ) from exc
+            raise
+
+
+class GDriveBufferedWriter(io.IOBase):
+    def __init__(self, fs, path):
+        self.fs = fs
+        self.path = path
+        self.buffer = io.BytesIO()
+        self._closed = False
+
+    def write(self, *args, **kwargs):
+        self.buffer.write(*args, **kwargs)
+
+    def readable(self):
+        return False
+
+    def writable(self):
+        return not self.readable()
+
+    def flush(self):
+        self.buffer.flush()
+        try:
+            self.fs.upload_fobj(self.buffer, self.path)
```

Comment:
just to double check - we need this to be append semantics, right? does it work this way?

Reply:
> we need this to be append semantics, right? does it work this way?

No. This is just a simple wrapper around `upload_fobj()` that buffers all your `write()` calls and dispatches them at once. Even after `flush()` we forcefully `close` the file so that you can not make any changes that might lead you to expected `append()` like semantics.

Reply:
I see. A few questions here:

- should we test `self._closed` in write and raise if it's `True`? It's not a very common behavior to close on `flush` right?
- `self.buffer.flush()` do we need this?
- what will be happening on gdrive_retry? do we handle this at all in the place where we depend on this? (retries are important usually for tests to be stable)

Reply:
> ? should we test self._closed in write and raise if it's True? It's not a very common behavior to close on flush right?

`buffer.flush()` actually checks for that, so that we don't have to manage individual states

> self.buffer.flush() do we need this?

answered above

> what will be happening on gdrive_retry? do we handle this at all in the place where we depend on this? (retries are important usually for tests to be stable)

`upload_fobj()` is retried.

Reply:
> buffer.flush() actually checks for that,

could clarify please? (I'm not sure we are on the same page :) )

> upload_fobj() is retried.

yep, I see, `gdrive_upload_fobj` is retried inside the `upload_fobj`

Reply:
> could clarify please? (I'm not sure we are on the same page :) )

When we close our wrapper file (via `.close()`), we also close the buffer (we call `buffer.close()). And if we try to call `buffer.flush()` after we close the file (and due to that, the buffer itself) it will raise the proper `I/O operation on closed file` error.

Reply:
hmm ... it looks I am missing something still, sorry :)

My concern was about the following workflow:

```python
writer.flush() # it closes itself which is not expected (usually)
writer.write("more stuff") # how does it behave now? should we signal that we are trying to write to a closed object?
```

Reply:
You can't do that with the current approach. We only allow `flush()`ing once, and after the `flush()` we close the file (which is a bit unorthodox, though was the easiest route to emulate the behavior since generally you only do it once during the `close()` itself so you don't actually need the file to be opened for the rest of the flow).

Reply:
Yep, so going back to the initial question "should we signal that we are trying to write to a closed object?" ?

Reply:
I think the answer is that, we already do at the first line of this function. 
```py
>>> import io
>>> buf = io.BytesIO()
>>> buf.write(b"hey") # first write data
3
>>> buf.flush(); buf.close() # then flush + close the file
>>> buf.flush() # try flushing again
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: I/O operation on closed file.
```

When you write stuff, and then call `flush()` of our wrapper (which also calls `close()` as I stated), then the next flush automatically is going to signal that since the first line is `buf.flush()` which would raise the appropriate error.